### PR TITLE
Vulkan: Workaround ARM Mali depth hardware bug. Fixes #11937

### LINKED ISSUE #11937: Mali hardware bug: Image corruption in Burnout on Android P on Galaxy S8+
Large graphical glitches whether or not a "sun readback" (lens flare effect) is happening, looks like tile buffer invalidation problems or similar - or a depth issue of sorts.

I've confirmed that it's not our regression by testing older builds.

Here's what a GE dump looks like (in game it's slightly different since it's overdrawing the last frame but it's still clearly wrong):

Screenshot_20190806-164431_PPSSPP

And here's a download of the GE dump:

[ULUS10025_0001.zip](https://github.com/hrydgard/ppsspp/files/3472839/ULUS10025_0001.zip)

Tested on Galaxy S8+ (SM-G955F). Driver version "04000000".

### PR DESCRIPTION
**Pull request by hrydgard**
Fixes #11937.

When triangles coincide with the Z/W=1 plane in specific ways, triggered by Burnout Legends' sky for example, the depth buffer gets corrupted.

~~This is worked around here by slightly rescaling Z. This type of workaround is recommended by ARM driver engineers.~~

EDIT: Actually, by nudging Z when Z equals W.

Ugly but what can you do when the hardware is bugged, confirmed by contacting ARM. I've done quick tests on a number of games with no issues. 

Not quite sure why I didn't see this on some older drivers, maybe the driver was doing an ugly workaround that was later removed, or something.

I'm awaiting information about which GPU hardware versions this is needed for, in the meantime let's do it on all Malis.

### CODE REVIEW

**Comment by unknownbrackets** [comment]
Hm.  When using GPU_SUPPORTS_ACCURATE_DEPTH, we already centrally scale depth, but using a matrix.  Is the Z=1 problem only before transform?  Easily could be confirmed by testing the `GPU_SCALE_DEPTH_FROM_24BIT_TO_16BIT` flag.

Does pixel rounding still work this way (like Phantasy Star Portable text)?

It's just that the vertex shader can't output 1.0f for z?

-[Unknown]

**Comment by hrydgard** [comment]
Waiting for information about that from ARM, but also I really need to write some tests to figure out the actual behaviors of the bug that matters for us. 

This change is just intended to be "bash it into working well enough for 1.9.0". 

But yeah I'll try that flag. Maybe we can just force it on instead, it doesn't have a big perf hit.

**Comment by Emulatorer** [comment]
@hrydgard Sir will you only enable this on all Mali-GPU that supports Vulkan api?
https://developer.arm.com/solutions/graphics/apis/vulkan/vulkan-drivers

Sorry for asking 😅

**Comment by hrydgard** [comment]
Very strangely, enabling vertex depth rounding breaks the graphics even more, almost nothing draws.

Pixel depth rounding doesn't make a difference.

I'll dig a little more...

**Comment by hrydgard** [comment]
When drawing with GPU_SCALE_DEPTH_FROM_24BIT_TO_16BIT in this game, the depth range is curiously large for the main scene draw, 0.00000756991 to 0.9999992 instead of 0.498047 to 0.50193 in the other passes, but seems the proj matrix makes up for that.

@unknownbrackets btw, unrelated since it doesn't happen in this game, but I think we may have a problem with clear depth. When we automatically convert draws to clears in DrawEngine (SW_CLEAR) we do not transform the depth according to the currently enabled depth scaling method...

File: GPU/Vulkan/GPU_Vulkan.cpp [hunk 1]
```diff
@@ -205,11 +205,16 @@ void GPU_Vulkan::CheckGPUFeatures() {
 		if (!PSP_CoreParameter().compat.flags().DisableAccurateDepth || driverTooOld) {
 			features |= GPU_SUPPORTS_ACCURATE_DEPTH;
 		}
+		// These GPUs (up to some certain hardware version?) has a bug where draws where gl_Position.w == .z
+		// corrupt the depth buffer. This is easily worked around by simply scaling Z down a tiny bit when this case
+		// is detected. See: https://github.com/hrydgard/ppsspp/issues/11937
+		features |= GPU_NEEDS_Z_EQUAL_W_HACK;
```
Comment (unknownbrackets) [inline GPU/Vulkan/GPU_Vulkan.cpp]: I wonder if these driver version checks ought to use the bugs_.Infest path.  Doesn't matter much, though.

-[Unknown]

File: GPU/Vulkan/DrawEngineVulkan.cpp [hunk 2]
```diff
@@ -962,6 +962,7 @@ void DrawEngineVulkan::DoFlush() {
 
 			// We let the framebuffer manager handle the clear. It can use renderpasses to optimize on tilers.
 			// If non-buffered though, it'll just do a plain clear.
+			// TODO: Don't we need to rescale result.depth if we use GPU_SCALE_DEPTH_FROM_24_TO_16_BIT and similar?
```
Comment (unknownbrackets) [inline GPU/Vulkan/DrawEngineVulkan.cpp]: As noted in the other issue, `result.depth` should already be the scaled value.

-[Unknown]

**Review by unknownbrackets — APPROVED** [review]
I wonder if GL shares this hardware bug and the driver works around it somehow...

-[Unknown]

File: GPU/Vulkan/GPU_Vulkan.cpp (hunk 1 above)
Comment (hrydgard) [reply GPU/Vulkan/GPU_Vulkan.cpp]: Probably. There are more things that should be on that path though, someday I'll clean it all up...

File: GPU/Vulkan/DrawEngineVulkan.cpp (hunk 2 above)
Comment (hrydgard) [reply GPU/Vulkan/DrawEngineVulkan.cpp]: Oh duh.

**Pull request merged**